feat: 초대 화면 UI 구현 (#40)

jus0k opened the pull request:
### Short description

#11 의 하위 과제인 초대 화면 UI 구현

### Proposed changes

- theme palette, typography 추가
- Header 공용 컴포넌트 추가
- Button 공용 컴포넌트 추가
- Layout 공용 컴포넌트 추가
- DotsIndicator (현재 내 화면이 어떤 단계에 있는지 표시) 공용 컴포넌트 추가
- 초대 페이지 제목 영역 추가
- 초대 페이지 InviteButton 컴포넌트 추가
- 이미지 에셋 (카카오톡, 링크 아이콘) 추가

### Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionally)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Test
- [ ] Documentation
- [ ] Refactoring

### How to test (Optional)

- 제플린과 구현된 UI 비교
- 이전 버튼 동작 확인
- 나중에 하기 버튼 동작 확인

### Reference (Optional)

- Fixes #40

Commit:
feat: 초대 화면 생성 및 Header, Button 공용 컴포넌트 추가 (#40)

- Invitation 컴포넌트 추가 및 라우트 연결
- Header 공용 컴포넌트 추가
- Button 공용 컴포넌트 추가

chayeoi commented:
master 브랜치와 sync 맞춰주시기 바랍니다~ 😄 

1. master 브랜치로 checkout 후 `git pull origin master --rerbase`
2. feature/11-invitation 브랜치로 checkout 후 `git rebase master`
3. 만약 2번 과정에서 conflict가 발생할 시, conflict가 발생한 부분을 해결하고 `git rebase --continue`를 차례차례 실행

jus0k commented:
> master 브랜치와 sync 맞춰주시기 바랍니다~ 😄
> 
> 1. master 브랜치로 checkout 후 `git pull origin master --rerbase`
> 2. feature/11-invitation 브랜치로 checkout 후 `git rebase master`
> 3. 만약 2번 과정에서 conflict가 발생할 시, conflict가 발생한 부분을 해결하고 `git rebase --continue`를 차례차례 실행

<img width="750" alt="스크린샷 2019-12-02 오후 10 51 05" src="https://user-images.githubusercontent.com/43605468/69964785-5b382b80-1556-11ea-95e6-5500e5310ee4.png">
동일하게 진행하고 conflict 없어서 다시 push 했는데 이런 오류가 나네요 ㅜㅜ

chayeoi commented on src/views/Invitation/Invitation.tsx:
```diff
@@ -12,6 +14,18 @@ const Invitation: React.FC = () => {
+
+  const handleLinkButton = (): void => {
+    console.log('link clicked');
+  };
```
로그를 언젠가는 지울거긴 해도, 어짜피 지울 코드이기 때문에 커밋 히스토리에 로그 출력 코드는 아예 남기지 않는게 깔끔할 것 같습니다!
181585d5690a24dddfa285043c27d274653e8f01 커밋에서 위 로그가 추가되었는데, interactive rebase 사용하셔서 변경하신 다음 `git push origin feature/11-invitation --force-with-lease` 해주시면 됩니다.

jus0k replied:
interactive rebase를 처음 봐서 어떻게 해야되는지 감이 안오네요 ㅜㅜ 해당 커밋에 변경사항이 많은데 `git rebase -i [HEAD~]` 후에 해당 커밋을 `edit` 하고  `git rebase --continue` 를 하면 되는건가요?

chayeoi replied:
맞습니다 정확합니다!

chayeoi commented on src/views/Invitation/components/InviteButton.tsx:
```diff
@@ -0,0 +1,43 @@
+}
+
+const InviteButton: React.FC<InviteButtonProps> = ({ onClick, label, icon, ...props }) => (
+  // eslint-disable-next-line
```
이 부분에 eslint disable 규칙이 어떤 점때문에 필요한가요~? 😄

jus0k replied:
...props 때문에 필요합니다!

chayeoi replied:
이 부분에 관한 eslint 규칙을 off하는 게 좋겠네요~

chayeoi commented on src/views/Invitation/components/InviteButton.tsx:
```diff
@@ -0,0 +1,43 @@
+  icon: string;
+}
+
+const InviteButton: React.FC<InviteButtonProps> = ({ onClick, label, icon, ...props }) => (
```
사용되지 않는 props에 대해 rest 오퍼레이터를 사용할 때, `props`라는 이름은 간혹 render prop 기법을 사용할 때 충돌하는 경향이 있어서 개인적으로 저는 `otherProps`라는 이름을 선호하는데요. `otherProps`로 통일하는 건 어떠신가요~? 😄

jus0k replied:
좋습니다!

chayeoi commented on src/views/assets/index.ts:
```diff
@@ -0,0 +1,2 @@
+export { default as kakaoIcon } from './ic_kakao.svg';
```
asset 네이밍에 관한 부분도 모두 통일했으면 하는데요~ 단어 간에 구분을 지을 때 언더스코어(`_`) 대신 하이픈(`-`)을 사용하면 어떨까 합니다. 언더스코어는 단어를 구분하는 역할이 아니라 단어의 한 부분으로 인식되기 때문인데 어떠신가요~? 😄

jus0k replied:
에셋은 모두 언더스코어를 사용한다고 어디서 들은 것 같아 변경했는데 그렇게 하겠습니다!

chayeoi replied:
[이런 글](https://medium.com/@AkhilDad/a-designers-guide-for-naming-android-assets-f790359d11e5)이 있긴 하네요. 우선 한 번 읽어볼게요!

chayeoi commented on src/views/components/Button/Button.tsx:
```diff
@@ -7,8 +7,9 @@ interface ButtonProps {
 
-const Button: React.FC<ButtonProps> = ({ onClick, label }) => (
-  <button type="button" css={s.button} onClick={onClick}>
+const Button: React.FC<ButtonProps> = ({ onClick, label, ...props }) => (
```
`onClick` prop도 `button` 요소에 같은 이름으로 전달되고 있는데 레스트(`...`) props에 포함시켜서 한 번에 내려줘도 괜찮을 듯합니다~ 혹시 이렇게 하면 안되는 이유가 있는건가요?

jus0k replied:
레스트 props는 button의 style 레스팅 적용과 관련된 부분이라 onClick과는 분리해서 사용하는 것이 좋지 않을까 했습니다!

chayeoi commented:
> > master 브랜치와 sync 맞춰주시기 바랍니다~ 😄
> > 
> > 1. master 브랜치로 checkout 후 `git pull origin master --rerbase`
> > 2. feature/11-invitation 브랜치로 checkout 후 `git rebase master`
> > 3. 만약 2번 과정에서 conflict가 발생할 시, conflict가 발생한 부분을 해결하고 `git rebase --continue`를 차례차례 실행
> 
> <img alt="스크린샷 2019-12-02 오후 10 51 05" width="750" src="https://user-images.githubusercontent.com/43605468/69964785-5b382b80-1556-11ea-95e6-5500e5310ee4.png">
> 
> 동일하게 진행하고 conflict 없어서 다시 push 했는데 이런 오류가 나네요 ㅜㅜ

`git push origin feature/11-invitation --force-with-lease`로 해주셔야 합니다~

chayeoi commented:
0f8a60205aa60b4c6db0cf3214d8e97495876fa0 커밋에 body가 비어있습니다. 정확히 어떤 목적으로 무엇을 변경했는지 내용을 채워주시면 좋을 것 같습니다.
또한 커밋 메세지 내용은 "헤더 공용 컴포넌트 추가"인데, 실제 코드에는 헤더 컴포넌트 추가 뿐 아니라 `Invitation` 페이지를 추가하는 등 다른 작업들이 한 커밋에 함께 들어가있네요~ 개별 커밋으로 분리하거나 해당 커밋에서 진행한 작업을 더 명확히 드러낼 수 있는 내용으로 변경 부탁드립니다! 제가 보기로는 이 커밋에서 진행한 작업은 파트너 초대 화면의 기본 레이아웃을 잡는 작업으로 보여요~

jus0k commented:
> [0f8a602](https://github.com/mash-up-kr/luvket/commit/0f8a60205aa60b4c6db0cf3214d8e97495876fa0) 커밋에 body가 비어있습니다. 정확히 어떤 목적으로 무엇을 변경했는지 내용을 채워주시면 좋을 것 같습니다.
> 또한 커밋 메세지 내용은 "헤더 공용 컴포넌트 추가"인데, 실제 코드에는 헤더 컴포넌트 추가 뿐 아니라 `Invitation` 페이지를 추가하는 등 다른 작업들이 한 커밋에 함께 들어가있네요~ 개별 커밋으로 분리하거나 해당 커밋에서 진행한 작업을 더 명확히 드러낼 수 있는 내용으로 변경 부탁드립니다! 제가 보기로는 이 커밋에서 진행한 작업은 파트너 초대 화면의 기본 레이아웃을 잡는 작업으로 보여요~

네 시도해보겠습니다!

chayeoi commented:
0f8a60205aa60b4c6db0cf3214d8e97495876fa0 커밋에 body가 비어있습니다. 정확히 어떤 목적으로 무엇을 변경했는지 내용을 채워주시면 좋을 것 같습니다.
또한 커밋 메세지 내용은 "헤더 공용 컴포넌트 추가"인데, 실제 코드에는 헤더 컴포넌트 추가 뿐 아니라 `Invitation` 페이지를 추가하는 등 다른 작업들이 한 커밋에 함께 들어가있네요~ 개별 커밋으로 분리하거나 해당 커밋에서 진행한 작업을 더 명확히 드러낼 수 있는 내용으로 변경 부탁드립니다! 제가 보기로는 이 커밋에서 진행한 작업은 파트너 초대 화면의 기본 레이아웃을 잡는 작업으로 보여요~

631a87e237f7545c238d799b49b04bf7364953ac 커밋에서 `master 브랜치 변경 사항 적용`이라는 한 줄 내용만 봐서는 어떤 작업을 했는지 이해하기 어려운 것 같습니다. 무슨 작업을 했는지 드러나도록 변경해주면 좋을 것 같습니다~ 또한 body 메세지 내용 중에 'chore:'로 시작하는 부분이 있는데, 아마도 40c4bbf1d8df4a46ee16608e78ceb134777ce596 커밋 메세지의 내용을 그대로 가져온 걸로 보입니다~
커밋 메세지 내용을 그대로 넣기보다는 위 커밋 해시를 메세지에 포함시켜서 클릭했을 때 직접 이동할 수 있도록 연결시켜주는 게 더 좋을 것 같아요~

chayeoi commented on src/views/Invitation/Invitation.tsx:
```diff
@@ -0,0 +1,23 @@
+      <Header>
+        <Button onClick={routePrevPage} label="이전" />
+      </Header>
+      <div>Invitation</div>
```
18행 `div`는 dd791b428d516fdab759644415f3dd6701588c07 에서 바로 삭제된 코드이기 때문에 커밋 히스토리에 남기지 않는 게 더 깔끔할 것 같습니다~

jus0k replied:
넵 알겠습니다!

chayeoi commented on src/views/Invitation/index.ts:
```diff
@@ -0,0 +1,5 @@
+import React from 'react';
+
+const Invitation = React.lazy((): Promise<{ default: React.ComponentType }> => import('./Invitation'));
```
제가 예시로 넣어놓은 `Home` 컴포넌트에는 prefetch를 빼먹었는데, 이 부분에 청크 파일 이름과 prefetch를 적용해주시면 좋을 것 같습니다~

```typescript
const Invitation: React.FC = React.lazy((): Promise<{ default: React.ComponentType }> => import(/* webpackChunkName: 'Invitation', webpackPrefetch: true */ './Invitation'));
```

jus0k replied:
넵 반영하겠습니다~ Home에도 해줘야 하나요?

chayeoi replied:
`Home`은 현재 커밋의 스코프가 아니고 이미 `master`에 적용된 상태여서 개별 커밋으로 남기는 게 좋을듯 합니다. 지금은 `Invitation`에만 적용해주세요!

jus0k replied:
넷 알겠습니다!

chayeoi commented on src/views/components/Button/Button.tsx:
```diff
@@ -1,99 +1,27 @@
+  <button type="button" css={s.button} onClick={onClick}>
+    {label}
   </button>
 );
```
1. 버튼 텍스트는 `children` prop으로 받도록 하는게 더 자연스러울 것 같습니다.
2. `button` 요소에 `type` 속성을 지정하지 않을 경우 기본값은 `submit`으로 지정되는데요. 현재는 `button` 타입으로 static하게 지정되어있을 뿐더러 `Button` 컴포넌트를 사용하는 시점에 `type` 속성을 명시적으로 전달할 수도 없어서 리팩토링이 필요해보입니다. 아래처럼 변경하면 좋을 것 같아요~ 😄 

```tsx
const Button: React.FC = (props) => <button css={s.button} {...props} />;

jus0k replied:
위에 처럼 했을 때 `Missing an explicit type attribute for button eslint(react/button-has-type)` 라는 에러가 나는데 이 린트 설정도 꺼도 괜찮은가요?

jus0k replied:
<img width="610" alt="스크린샷 2019-12-04 오후 9 39 19" src="https://user-images.githubusercontent.com/43605468/70143324-9d8c7480-16de-11ea-9b5b-c361ad37c3a1.png">
Button 컴포넌트를 사용할 때도 onClick 을 넣었을 때 위와 같은 에러가 나네요!

chayeoi replied:
`react/button-has-type`은 우선 off하도록 하고 마지막에 남겨주신 에러에 대해 좀 더 자세히 설명해주실수 있나요?

jus0k replied:
<img width="516" alt="스크린샷 2019-12-05 오후 5 59 23" src="https://user-images.githubusercontent.com/43605468/70219924-07615880-1789-11ea-9361-d6bdee4f2f13.png">
위와 같이 Button 컴포넌트를 다른 컴포넌트에 불러와서 onClick 이벤트를 할당했을 때 위와 같은 에러가 나는 것 같습니다!

jus0k replied:
```javascript 
const Button: React.FC = (props) => <button css={s.button} {...props} />;
```
위에서 Button 컴포넌트 props에 대한 Type 정의가 안되어 있어서 그런걸까요?

jus0k replied:
<img width="658" alt="스크린샷 2019-12-05 오후 6 37 16" src="https://user-images.githubusercontent.com/43605468/70223069-62e21500-178e-11ea-8abd-593b6c618fc8.png">
이렇게 하니까 해결되었습니다! 다만 빈 interface를 만들지 말라는 eslint 규칙이 있어서 일단은 off 하고 진행해도 괜찮을까요?

chayeoi replied:
넵 좋습니다!

chayeoi commented on src/views/components/Button/Button.tsx:
```diff
@@ -1,99 +1,27 @@
 /** @jsx jsx */
```
이 부분도 현재 PR에 추가된 b24919d760f5ddc2295eed786b24bf0e79f43c98 에서 제거되는 코드기 때문에, 현재 커밋에서 해당 코드를 넣어주면 불필요한 커밋을 줄일 수 있을 것 같습니다~

jus0k replied:
넵 알겠습니다~

chayeoi commented on src/views/components/Header/Header.tsx:
```diff
@@ -0,0 +1,25 @@
+  children: React.ReactNode;
+}
+
+const Header: React.FC<HeaderProps> = ({ children }) => <div css={s.wrapper}>{children}</div>;
```
1. 지금처럼 한 컴포넌트가 단일 요소만 갖는 경우에는 인자로 들어온 props를 스프레드 연산자로 모두 내려주도록 하는게 실수를 방지할 수 있을 것 같습니다~
2. `Header`라는 용도에 맞게 `div` 대신 `header` 요소를 사용하는 게 더 적절할 것 같습니다.

```tsx
const Header: React.FC<HeaderProps> = (props) => <div css={s.wrapper} {...props} />;
```

jus0k replied:
넵 반영하겠습니다~

chayeoi commented on src/views/styles/global.ts:
```diff
@@ -3,6 +3,9 @@ import emotionReset from 'emotion-reset';
 
 const global = css`
   ${emotionReset}
+  * {
```
가상 요소에도 해당 속성을 적용해주시면 좋을 것 같습니다.

```css
*, *::before, *::after {
  box-sizing: border-box;
}
```

jus0k replied:
오호 넵 적용하겠습니다!

chayeoi commented on src/views/styles/theme.ts:
```diff
@@ -1,7 +1,14 @@
-  palette: {},
+  palette: {
+    gray: '#d8d8d8',
+  },
```
dd791b428d516fdab759644415f3dd6701588c07 커밋에서 `gray` 인터페이스가 변경되었는데, 이 부분에서 미리 그 형식에 맞게 변경해주시면 히스토리가 더 깔끔할 것 같습니다~

jus0k replied:
```javascript
  palette: {
    gray: {
      100: '#d8d8d8',
    },
```
다음 커밋에서 추가되는 500은 추가하지 말고 그냥 구조만 이렇게 바꾸면 되는거 맞나요?

chayeoi replied:
맞습니다~

chayeoi commented on src/views/Invitation/Invitation.tsx:
```diff
@@ -0,0 +1,23 @@
+import React from 'react';
+import { useHistory } from 'react-router';
```
`useHistory`는 `react-router-dom`으로부터 import해야 합니다. dd791b428d516fdab759644415f3dd6701588c07 커밋에서 `react-router-dom`으로 변경하는 부분이 있는데, 변경하는 부분을 커밋에 포함시키지 말고 이곳에서 `react-router-dom`으로부터 가져오도록 하면 더 좋을 것 같습니다~

jus0k replied:
넵! 반영하겠습니닷

chayeoi commented on src/views/Invitation/Invitation.tsx:
```diff
@@ -1,7 +1,10 @@
+/** @jsx jsx */
```
이 부분도 어짜피 뒤에서 지워질 부분이기 때문에 제거해주셔도 좋을 것 같습니다~

jus0k replied:
넵 알겠습니다~

chayeoi commented on src/views/Invitation/Invitation.tsx:
```diff
@@ -11,13 +14,31 @@ const Invitation: React.FC = () => {
-      <div>Invitation</div>
-    </>
+      <Layout>
+        <h1 css={s.title}>초대하기</h1>
```
`h1`은 앱 전체 서비스의 제목을 나타내기 위한 요소로 단 한번만 사용되는데요~ 페이지에 대한 제목이므로 `h2` 요소를 사용하는 게 더 적절할 것 같습니다.

jus0k replied:
반영하겠습니다~

chayeoi commented on src/views/components/Layout/Layout.tsx:
```diff
@@ -0,0 +1,17 @@
+/** @jsx jsx */
```
이 부분도 지워도 될 것 같습니다~

chayeoi commented on src/views/components/Layout/Layout.tsx:
```diff
@@ -0,0 +1,17 @@
+  children: React.ReactNode;
+}
+
+const Layout: React.FC<LayoutProps> = ({ children }) => <div css={s.wrapper}>{children}</div>;
```
이 부분도 전체 `props`를 spread하도록 하면 좋을 것 같습니다~

```tsx
const Layout: React.FC<LayoutProps> = (props) => <div css={s.wrapper} {...props} />;

```

chayeoi commented on src/views/Invitation/Invitation.tsx:
```diff
@@ -0,0 +1,23 @@
+  };
+
+  return (
+    <>
```
`Fragment` 요소가 dd791b428d516fdab759644415f3dd6701588c07 커밋에서 `React.Fragment`로 변경되었고, b24919d760f5ddc2295eed786b24bf0e79f43c98 커밋에서 다시 빈 태그로 변경되고 있습니다. 이 부분은 18행을 지워주면 `Fragment`를 사용할 필요가 없기 때문에 이번 커밋에서 아예 지워주셔도 좋을 듯합니다~

chayeoi commented on src/views/Invitation/Invitation.tsx:
```diff
@@ -0,0 +1,23 @@
+
+  const routePrevPage = (): void => {
+    history.goBack();
+  };
```
`history.goBack`을 16행의 `onClick` prop에 바로 할당하거나 `handleClick`으로 네이밍하는건 어떨까요~? 이벤트 핸들러 네이밍 규칙에 관해 다음 번에 한 번 논의해보면 좋을 것 같습니다!

chayeoi commented on src/views/components/Button/Button.tsx:
```diff
@@ -1,5 +1,4 @@
-/** @jsx jsx */
-import { css, jsx } from '@emotion/core';
+import { css, SerializedStyles } from '@emotion/core';
```
`SerialziedStyles`를 추가하는 부분을 `Button` 컴포넌트를 작업했던 커밋으로 옮기고, 이 커밋은 제거해도 좋을 것 같습니다~

chayeoi commented on src/views/Invitation/Invitation.tsx:
```diff
@@ -12,6 +14,14 @@ const Invitation: React.FC = () => {
+
+  const handleKakaoButton = (): void => {};
+
+  const handleLinkButton = (): void => {};
```
아직 저희끼리 정한 이벤트 핸들러 네이밍 규칙이 없는데, 우선은 [여기 링크](https://jaketrent.com/post/naming-event-handlers-react/) 내용을 참고해주시면 좋을 것 같습니다~ 현재는 요소에 대한 이름만 정의되어 있고 액션에 대한 정의 이름에 포함되어있지 않네요~

chayeoi commented on src/views/Invitation/Invitation.tsx:
```diff
@@ -23,6 +33,22 @@ const Invitation: React.FC = () => {
+            icon={kakaoIcon}
+            label="카카오톡으로 초대하기"
+            onClick={handleKakaoButton}
+            css={{ marginBottom: '10px' }}
```
41행과 47행 부분에만 스타일 객체를 인라인으로 정의하신 이유가 있나요~? 특별한 이유가 없다면 스타일 객체로 분리하는 게 깔끔할 것 같습니다~

chayeoi commented:
각 커밋의 메세지 한줄 내용 마지막에 이슈 번호 (#40)을 남겨주면 좋을 것 같아요~ 😄

Commit:
feat: 초대 화면 제목 영역 추가 (#40)

- 화면 내 제목 및 설명 영역 추가
- Layout 공용 컴포넌트 추가
- typography theme 정의 및 css 작성

Signed-off-by: jus0k <jutjd115@gmail.com>

Commit:
feat: 초대 화면 버튼 영역, Dots Indicator UI 추가 (#40)

- 카카오톡 공유, 링크 복사하기 버튼 UI 구현
- 텍스트 버튼 UI 구현
- 하단 Dots Indicator UI 구현
- UI 구현을 위한 theme 추가

Signed-off-by: jus0k <jutjd115@gmail.com>

chayeoi commented on src/views/styles/theme.ts:
```diff
@@ -1,7 +1,16 @@
-  palette: {},
+  palette: {
+    gray: {
+      100: 'd8d8d8',
```
f4e3666214a69e106265aec13f86bd080ec82859 커밋에서 앞에 `#` 기호 붙이는 부분이 추가되고 있는데, 이 커밋에서 넣는 걸로 수정하면 좋을 것 같습니다~

jus0k replied:
넵 반영했습니다!
